### PR TITLE
Separate mouse and touch events

### PR DESCRIPTION
**Pull request by juanjoDiaz**
This PR is an initial step towards #1267 
It separates the mouse event handling from the touch event handling.

A separate step could be to implement multiple Touch modules for different interaction styles a.k.a convert touch events to mouse events in different ways.
For example, one that would work for me would be something like:

| Touch action | Click event |
|------|------|
| Single Tap| Click|
| Long press | Right-click (contextmenu or click event)|
| Press and then drag | Click and drag (mousemove event) |
| Two-finger press and drag | Pan (Wheel event)|
| Two-finger pinch | Zoom (Not sure of the events here) |

A library like hammer.js would make such implementations very easy.

### CODE REVIEW

File: core/rfb.js [hunk 1]
```diff
@@ -177,8 +179,14 @@ export default class RFB extends EventTargetMixin {
         this._keyboard.onkeyevent = this._handleKeyEvent.bind(this);
 
         this._mouse = new Mouse(this._canvas);
-        this._mouse.onmousebutton = this._handleMouseButton.bind(this);
-        this._mouse.onmousemove = this._handleMouseMove.bind(this);
+        this._mouse.onmousebutton = this._handlePointerPress.bind(this);
+        this._mouse.onmousemove = this._handlePointerMove.bind(this);
+
+        if (isTouchDevice) {
+            this._touch = new Touch(this._canvas, this._mouse);
```
Comment (samhed) [inline core/rfb.js]: The second argument to the constructor looks like it should be removed.

File: core/rfb.js [hunk 2]
```diff
@@ -177,8 +179,14 @@ export default class RFB extends EventTargetMixin {
...
+            this._mouse.ontouch = this._handlePointerPress.bind(this);
+            this._mouse.ontouchmove = this._handlePointerMove.bind(this);
```
Comment (samhed) [inline core/rfb.js]: I believe these two should be `this._touch.ontouch` and `this._touch.ontouchmove`

File: docs/API-internal.md [hunk 3]
```diff
@@ -58,29 +58,52 @@ callback event name, and the callback function.
...
+| name   | parameters | description
+| ------ | ---------- | ------------
+| grab   | ()         | Begin capturing touch events
+| ungrab | ()         | Stop capturing touch events
+
+### 2.2.2 Callbacks
+
+| name          | parameters          | description
+| ------------- | ------------------- | ------------
+| ontouch       | (x, y, down, bmask) | Handler for touch even (as button click/release)
```
Comment (samhed) [inline docs/API-internal.md]: "event"

File: core/rfb.js [hunk 4]
```diff
@@ -264,17 +272,20 @@ export default class RFB extends EventTargetMixin {
             if (viewOnly) {
                 this._keyboard.ungrab();
                 this._mouse.ungrab();
+                if (this._touch) { this._touch.ungrab(); }
             } else {
                 this._keyboard.grab();
                 this._mouse.grab();
+                this._mouse.grab();
```
Comment (samhed) [inline core/rfb.js]: Looks like one `_mouse.grab()` too many

**Comment by samhed** [comment]
And btw regarding hammerjs it would have been nice, but that project seems dead seeing as the latest release was more than 3 years ago and the maintainers don't seem to be handling issues or pull requests anymore.

If anyone knows of a good alternative library please tell us.

**Comment by juanjoDiaz** [comment]
Fixed your comments.
I wanted to check if this is wanted before going further since I've had plenty of PRs rejected in this project 🙂
Some testing is still needed for sure. I'd be testing more on Monday.

Regarding Hammer.js, I know that it's not the best-maintained project but it is still the de-facto library for touch support on websites and it has over 1 million weekly downloads in npm. So not sure if there is a better alternative but I couldn't find any and it seems unlikely that there is one.

**Comment by CendioOssman** [comment]
:+1: to the principle of splitting things. Haven't checked the code yet.

FYI, we (Cendio) are planning on getting some gestures up and running in the coming months. This would probably be a good base for that.

**Comment by juanjoDiaz** [comment]
Is there anything left for me to do in this PR?

**Comment by samhed** [comment]
Looks like all my code-comments are addressed. How well tested is this?

**Comment by juanjoDiaz** [comment]
I've been using noVNC.html on a tablet with no problem.

Can you think of any particular thing that I should test?

**Comment by CendioOssman** [comment]
I did a quick read of the code and it looks good AFAICT.

But I realized a big issue; the browser are moving away from the mouse and touch events in favour of common pointer events. Is this split here then perhaps counter productive as we have to have common code once we handle pointer events?

**Comment by juanjoDiaz** [comment]
I don't think that it's counter-productive.
We have a handler for a mouse event, a handler for touch events, and we could add a handler for Pointer events.

Then the logic would be:
```js
if (arePointerEventsSupported) {
  // use pointer events handler
} else {
  // use mouse events handler

  if (isTouchDevice) {
    // use touch events handler
  }
}
```

**Comment by CendioOssman** [comment]
I guess. No risk of a need for sharing code in the pointer handling and the other two though?

We will probably be looking at the pointer events soon-ish. Is it okay if we hold off on this PR a few weeks until we can get a better idea of how those work? I'd just like to confirm that we're not shuffling code back and forth needlessly and making git blame a pain.

**Comment by juanjoDiaz** [comment]
I don't think that there is any shared code between those 3 handlers since the shared parts are the actual event handlers which are set in the `rfb.js` file.
Also, if anything else should be shared it can be placed in a helper class that's composed into the handlers. That would provide a nice and clean separation of concerns imho.

I'm OK waiting a few weeks for this. Just let me know if there is anything that I can do to help.

When looking at pointer-events are you just looking at using them instead of mouse or touch events? Or are you planning to add better gesture support as you also commented above?

**Comment by CendioOssman** [comment]
> When looking at pointer-events are you just looking at using them instead of mouse or touch events? Or are you planning to add better gesture support as you also commented above?

The primary motivator is gesture support, which was originally intended to be built upon touch events. But as we're seeing issues with them, we will probably look at pointer events primarily now and keep mouse/touch support around only as much as is needed for legacy. I don't know what this means specifically until we've looked at pointer events more closely and evaluated which browsers support them.

**Comment by CendioNiko** [comment]
We have been in contact with upstream about pointer events, and we reached a decision not to migrate fully to pointer events as of now.  This is because they are ending support for 'ontouch' and createEvent('TouchEvent') etc. but they are seemingly keeping addEventListener(). So this PR is of great interest!

However when testing this code I encountered the following problem:
 * 'Touch and drag' events does not seem to get captured by our session anymore. I encountered this error on most platforms with touch. Tested on Chrome/Firefox on android, Chrome in Linux and Chrome/Edge on Windows 10.

Could you have a look at this?
If you don't have any touch device to try this out you can use touch emulation on Chrome.

**Comment by juanjoDiaz** [comment]
I'll try to take a look at this during this week.

What's not working is, for example, dragging icons on a desktop, right?

**Comment by CendioNiko** [comment]
Thank you!
Yeah, you should be able to see the problem by trying to drag an icon.

**Comment by samhed** [comment]
Replaced by changes made in #1414

**Pull request closed without merging**